hack/Dockerfile.buildproto

```
@@ -0,0 +1,36 @@
+#   Copyright The containerd Authors.
+
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+
+#       http://www.apache.org/licenses/LICENSE-2.0
+
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+ARG GO_VERSION=1.24
+
+FROM golang:${GO_VERSION}-bullseye AS builder
+ARG ARTIFACTS
+
+WORKDIR /go/src
+
+COPY . .
+
+RUN apt-get update && apt-get install -y unzip
+
+RUN --mount=type=cache,target=/go/pkg/mod/ \
+    make install-protoc-dependencies install-ttrpc-plugin install-wasm-plugin install-protoc
+
+RUN --mount=type=cache,target=/go/pkg/mod/ \
+    make build-proto
+
+RUN tar czf /artifacts.tgz ${ARTIFACTS}
```

i.e., these both could use a mount for the source. One thing to account for is that (unfortunately) the `make proto-build` defaults to writing to the source location, so it needs the build-context to be built writable (and combined with the `tar` step, as writable mounts are ephemeral);

```console
> [builder 5/6] RUN --mount=type=cache,target=/go/pkg/mod/     --mount=src=.,target=.     make build-proto:
0.317 Generating pkg/api/api.pb.go...
0.425 ./pkg/api/api.pb.go: Read-only file systemmake: *** [Makefile:181: pkg/api/api.pb.go] Error 1
```

Wondering if it'd be possible to make the proto-build accept a destination, which could possibly also avoid having to tar all the files to overwrite them (not just the generated files).



```suggestion
RUN --mount=type=cache,target=/go/pkg/mod/ \
    --mount=src=.,target=. \
    make install-protoc-dependencies install-ttrpc-plugin install-wasm-plugin install-protoc

RUN --mount=type=cache,target=/go/pkg/mod/ \
    --mount=src=.,target=.,rw=true \
    make build-proto && \
    tar czf /artifacts.tgz ${ARTIFACTS}
```

---

But maybe @crazy-max has some tricks up his sleeve; I recall some of the BuildKit "generated files" dockerfiles also synchronise files (but possibly in a more granular way).